pkg/scheduler/flavorassigner/flavorassigner.go

```diff
@@ -219,35 +331,46 @@ func (a *Assignment) findFlavorForCodepResources(
 		if match, err := selector.Match(&corev1.Node{ObjectMeta: metav1.ObjectMeta{Labels: flavor.NodeSelector}}); !match || err != nil {
 			if err != nil {
 				status.err = err
-				return nil, &status
+				return nil, status
 			}
 			status.append(fmt.Sprintf("flavor %s doesn't match with node affinity", flvLimit.Name))
 			continue
 		}
 
 		assignments := make(ResourceAssignment, len(requests))
+		worstFlavorMode := Fit
 		for name, val := range requests {
 			codepFlvLimit := cq.RequestableResources[name].Flavors[i]
 			// Check considering the flavor usage by previous pod sets.
-			borrow, s := fitsFlavorLimits(name, val+a.usage[name][flavor.Name], cq, &codepFlvLimit)
-			if s.IsError() {
-				return nil, s
-			}
-			if !s.IsSuccess() {
+			mode, borrow, s := fitsFlavorLimits(name, val+a.usage[name][flavor.Name], cq, &codepFlvLimit)
+			if s != nil {
 				status.reasons = append(status.reasons, s.reasons...)
+			}
+			if mode < worstFlavorMode {
+				worstFlavorMode = mode
+			}
+			if worstFlavorMode == NoFit {
+				// The flavor doesn't fit, no need to check other resources.
 				break
 			}
+
 			assignments[name] = &FlavorAssignment{
 				Name:   flavor.Name,
-				Mode:   CohortFit,
+				Mode:   mode,
 				borrow: borrow,
 			}
 		}
-		if len(assignments) == len(requests) {
-			return assignments, nil
+
+		if worstFlavorMode > bestAssignmentMode {
+			bestAssignment = assignments
+			bestAssignmentMode = worstFlavorMode
```

Comment:
this is a unintuitive, assigning the worst as the best :); perhaps rename `bestAssignmentMode` to `selectedMode`